tools/changelog-helper.sh

```diff
@@ -0,0 +1,283 @@
+    if [[ ${title_suggestion_in_pr} ]]; then
+        title=${title_suggestion_in_pr}
+        if [[ ${title_suggestion_in_pr} == "SKIP" ]]; then
+            return
+        fi
+    fi
+    echo -n "-> Missing PR #${id} (${title}, @${author})"
+    if [[ $ACTION != add-missing-entries ]]; then
+        echo
+        return
+    fi
+    echo ", adding new entry"
+    local new_entry=""
+    local lang=$(grep -oP 'Updated? \K(\S+)(?= app translations? for )' <<<"$title" || true)
+    if [[ $lang ]]; then
+        # Note: This creates a top-level entry for each language.
+        # group-entries can merge those to a single one.
+        local full_lang=${LANGS[$lang]-${lang}}
+        add_translation_pr "${lang}" "${author}" "${id}"
+        return
+    fi
+    new_entry=$(
+        echo "- ${title} (#${id})."
+        echo "  (contributed by @${author})"
+    )
+    local changelog_before=$(head -n $changelog_begin_position ChangeLog)
+    local changelog_after=$(tail -n +$(($changelog_begin_position+1)) ChangeLog)
+    (echo "$changelog_before"; echo; echo "$new_entry"; echo "$changelog_after") > ChangeLog
+}
+
+add_translation_pr() {
+    local lang="${1}"
+    local author="${2}"
+    local id="${3}"
+    local changelog_begin_position=$(grep -nP '^### .*\d+\.\d+\.\d+\b' ChangeLog | head -n1 | cut -d: -f1)
+    local changelog_prev_release_position=$(grep -nP '^### .*\d+\.\d+\.\d+\b' ChangeLog | head -n2 | tail -n1 | cut -d: -f1)
+
+    # Save everything before the actual release changelog content:
+    local changelog_header=$(head -n $changelog_begin_position ChangeLog)
+
+    # Save everything after the actual release changelog content:
+    local changelog_prev_releases=$(tail -n +$(($changelog_prev_release_position)) ChangeLog)
+
+    # Save the current release's changelog content:
+    local changelog=$(sed -rne '/^###.*'${target_release//./\.}'\b/,/^### '${prev_release//./\.}'\b/p' ChangeLog | tail -n +2 | head -n -1)
+    local changelog_orig="${changelog}"
+
+    # Is there an existing entry for this language already?
+    changelog=$(sed -re "s/^(  \* ${full_lang}, by .+ \(.*)\)/\1, #${id})/" <<<"${changelog}")
+    if [[ "${changelog}" == "${changelog_orig}" ]]; then
+        # No existing language entry. Check for an existing translation entry.
+        changelog=$(sed -re "s/^(- ${TRANSLATION_ENTRY_TEXT}.*)/\1\n  * ${full_lang}, by @${author} (#${id})/" <<<"${changelog}")
+        if [[ "${changelog}" == "${changelog_orig}" ]]; then
+            # No existing translation entry at all. Add a new one.
+            changelog="${changelog}$(
+                echo
+                echo
+                echo "- ${TRANSLATION_ENTRY_TEXT}"
+                echo "  * ${full_lang}, by @${author} (#${id})"
+			)"
+        else
+            # Existing translation entries, so sort them:
+            local changelog_before_translations=""
```

Comment:
Can sed edit "in-place" so we don‘t need to copy loads of data into a variable?

Reply:
`sed` does support in-place editing, but I don't think those features are sufficient to do what's needed here. Feel free to suggest an alternative, that's what I came up with in absence of any better approaches.

Reply:
Maybe include a FIXME comment explaining that this could be improved